src/replication.c

```diff
@@ -0,0 +1,1067 @@
+	if(rep_cmd->wrio_seq == replica->least_recvd + 1) {
+		while((rep_cmd = TAILQ_FIRST(&replica->waitq))) {
+			rcommq_ptr = rep_cmd->rcommq_ptr;
+			if(rep_cmd->ack_recvd == true) {
+				replica->least_recvd++;
+				MTX_UNLOCK(&replica->r_mtx);
+				rcommq_ptr->acks_recvd++;
+				luworker_id = rcommq_ptr->luworker_id;
+
+				rcommq_ptr->bitset &= ~(1 << replica->id);
+				if (rcommq_ptr->acks_recvd == spec->consistency_count) {
+					rcommq_ptr->status = 1;
+					//Transfer cmd from waitq to pendingq and wakeup luworker
+					MTX_LOCK(&spec->rcommonq_mtx);
+					TAILQ_REMOVE(&spec->rcommon_waitq, rcommq_ptr, wait_cmd_next);
+					if (rcommq_ptr->acks_recvd != spec->replica_count) {
+						TAILQ_INSERT_TAIL(&spec->rcommon_pendingq, rcommq_ptr, pending_cmd_next);
+					} else {
+						for (i=1; i < rcommq_ptr->iovcnt + 1; i++) {
+							xfree(rcommq_ptr->iov[i].iov_base);
+						}
+						rcommq_ptr->completed = 1;
+					}
+					MTX_UNLOCK(&spec->rcommonq_mtx);
+
+					MTX_LOCK(&spec->luworker_rmutex[luworker_id]);
+					pthread_cond_signal(&spec->luworker_rcond[luworker_id]);
+					MTX_UNLOCK(&spec->luworker_rmutex[luworker_id]);
+
+				} else if (rcommq_ptr->acks_recvd == spec->replica_count) {
+					MTX_LOCK(&spec->rcommonq_mtx);
+					TAILQ_REMOVE(&spec->rcommon_pendingq, rcommq_ptr, pending_cmd_next);
+					for (i=1; i < rcommq_ptr->iovcnt + 1; i++) {
+						xfree(rcommq_ptr->iov[i].iov_base);
+					}
+					if(rcommq_ptr->completed == 1) {
+						free(rcommq_ptr);
+						rcommq_ptr = NULL;
+					}
+					MTX_UNLOCK(&spec->rcommonq_mtx);
+				}
+				MTX_LOCK(&replica->r_mtx);
+				TAILQ_REMOVE(&replica->waitq, rep_cmd, rwait_cmd_next);
+				free(rep_cmd);
+				rep_cmd = NULL;
+			} else {
+				break;
+			}
+		}
+	}
+	MTX_UNLOCK(&replica->r_mtx);
+	return 0;
+}
+//Receive IO responses in this thread
+void *
+replica_receiver(void *arg) {
+	spec_t *spec = (spec_t *)arg;
+	replica_t *replica;
+	int i, epfd, event_count, replica_count = 0;
+	int64_t count = 0;
+	void *data;
+	bool read_io = false;
+	struct epoll_event event, *events;
+	size_t rsp_len = sizeof(zvol_io_hdr_t);
+	zvol_io_hdr_t *io_rsp = NULL;
+	//Create a new epoll epfd
+	epfd = epoll_create1(0);
+	events = calloc(MAXEVENTS, sizeof(event));
+	while(1) {
+		if(replica_count == 0) {
+			MTX_LOCK(&spec->rq_mtx);
+			if(spec->replica_count == 0) {
+				//Wait until at least one IO connection has been made to a registered replica
+				pthread_cond_wait(&spec->rq_cond, &spec->rq_mtx);
+			}
+			MTX_UNLOCK(&spec->rq_mtx);
+		}
+		//Check if number of available replicas have changed
+		if(replica_count != spec->replica_count) {
+			replica_count = update_replica_list(epfd, spec, replica_count);
+		}
+		//Wait for events on all iofds
+		event_count = epoll_wait(epfd, events, MAXEVENTS, -1);
+		for(i=0; i< event_count; i++) {
+			//Remove the replica from queue
+			if ((events[i].events & EPOLLERR) ||
+					(events[i].events & EPOLLHUP) ||
+					(!(events[i].events & EPOLLIN))) {
```

Comment:
EINTR should not be treated as error

Reply:
same with other places also..
actually, I have one doubt also.. do we get events only for those which are set in event.events? :)